network/peer/upgrader_test.go

```diff
@@ -0,0 +1,222 @@
+
+func TestBlockClientsWithIncorrectRSAKeys(t *testing.T) {
+	for _, testCase := range []struct {
+		description      string
+		genClientTLSCert func() tls.Certificate
+		expectedErr      error
+	}{
+		{
+			description: "Proper key size and private key - 2048",
+			genClientTLSCert: func() tls.Certificate {
+				privKey2048, err := rsa.GenerateKey(rand.Reader, 2048)
+				require.NoError(t, err)
+				clientCert2048 := makeTLSCert(t, privKey2048)
+				return clientCert2048
+			},
+		},
+		{
+			description: "Proper key size and private key - 4096",
+			genClientTLSCert: func() tls.Certificate {
+				privKey4096, err := rsa.GenerateKey(rand.Reader, 4096)
+				require.NoError(t, err)
+				clientCert4096 := makeTLSCert(t, privKey4096)
+				return clientCert4096
+			},
+		},
+		{
+			description: "Too big key",
+			genClientTLSCert: func() tls.Certificate {
+				block, _ := pem.Decode(fat8192BitRSAKey)
+				require.NotNil(t, block)
+				rsaFatKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+				require.NoError(t, err)
+				privKey8192 := rsaFatKey.(*rsa.PrivateKey)
+				// Sanity check - ensure privKey8192 is indeed an 8192 RSA key
+				require.Equal(t, 8192, privKey8192.N.BitLen())
+				clientCert8192 := makeTLSCert(t, privKey8192)
+				return clientCert8192
+			},
+			expectedErr: staking.ErrUnsupportedRSAModulusBitLen,
+		},
+		{
+			description: "Improper public exponent",
+			genClientTLSCert: func() tls.Certificate {
+				clientCertBad := makeTLSCert(t, nonStandardRSAKey(t))
+				return clientCertBad
+			},
+			expectedErr: staking.ErrUnsupportedRSAPublicExponent,
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			serverKey, err := rsa.GenerateKey(rand.Reader, 2048)
+			require.NoError(t, err)
+
+			serverCert := makeTLSCert(t, serverKey)
+
+			config := peer.TLSConfig(serverCert, nil)
+
+			c := prometheus.NewCounter(prometheus.CounterOpts{})
+
+			// Initialize upgrader with a mock that fails when it's incremented.
+			failOnIncrementCounter := &mockPrometheusCounter{
+				Counter: c,
+				t:       t,
+				onIncrement: func() {
+					require.FailNow(t, "should not have invoked")
+				},
+			}
+			upgrader := peer.NewTLSServerUpgrader(config, failOnIncrementCounter)
+
+			clientConfig := tls.Config{
+				ClientAuth:         tls.RequireAnyClientCert,
+				InsecureSkipVerify: true, //#nosec G402
+				MinVersion:         tls.VersionTLS13,
+				Certificates:       []tls.Certificate{testCase.genClientTLSCert()},
+			}
+
+			listener, err := net.Listen("tcp", "127.0.0.1:0")
+			require.NoError(t, err)
+			defer listener.Close()
+
+			var wg sync.WaitGroup
+			wg.Add(1)
+
+			go func() {
+				defer wg.Done()
+				conn, err := listener.Accept()
+				require.NoError(t, err)
+
+				_, _, _, err = upgrader.Upgrade(conn)
+
+				require.ErrorIs(t, err, testCase.expectedErr)
+			}()
+
+			conn, err := tls.Dial("tcp", listener.Addr().String(), &clientConfig)
+			require.NoError(t, err)
+
+			require.NoError(t, conn.Handshake())
```

Comment:
I'm a bit surprised that this passes, is this just because the client signature verification is performed last?

Reply:
In TLS1.3 we first negotiate the encryption, and do the authentication last. The client is the last party to send its authentication, and the implementation doesn't wait for the acknowledgement from the server, i guess for performance reasons.